src/engraving/rendering/engravingitempreviewpainter.cpp

```diff
@@ -0,0 +1,191 @@
+/*
+ * SPDX-License-Identifier: GPL-3.0-only
+ * MuseScore-Studio-CLA-applies
+ *
+ * MuseScore Studio
+ * Music Composition & Notation
+ *
+ * Copyright (C) 2024 MuseScore Limited
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+#include "engravingitempreviewpainter.h"
+
+#include "engraving/dom/actionicon.h"
+#include "engraving/style/defaultstyle.h"
+#include "engraving/dom/masterscore.h"
+
+using namespace mu::engraving::rendering;
+using namespace muse::draw;
+
+/// Paint an EngravingItem - include a staff if specified in the params.
+void EngravingItemPreviewPainter::paintPreview(mu::engraving::EngravingItem* element, PaintParams& params)
+{
+    IF_ASSERT_FAILED(element && params.painter) {
+        return;
+    }
+
+    Painter* painter = params.painter;
+    painter->setPen(params.color);
+
+    if (element->isActionIcon()) {
+        paintActionIcon(element, params);
+        return; // never draw staff for icon elements
+    }
+
+    PointF origin = params.rect.center(); // draw element at center of cell by default
+    if (params.drawStaff) {
+        const double topLinePos = paintStaff(params); // draw dummy staff lines onto rect.
+        origin.setY(topLinePos); // vertical position relative to staff instead of cell center.
+    }
+
+    painter->translate(origin);
+
+    painter->translate(params.xoffset * params.spatium, params.yoffset * params.spatium); // additional offset for element only
+
+    paintEngravingItem(element, params);
+}
+
+void EngravingItemPreviewPainter::paintSingleItem(mu::engraving::EngravingItem* element, PaintParams& params)
```

Comment:
One might argue that it's not ideal that not all fields of `params` are actually used by this method, which might create false expectations about what it does. But it's not really a big deal to me.